pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,394 @@
+func (e *EventedPLEG) updateGlobalCache() {
+	e.cache.UpdateTime(time.Now())
+}
+
+// Update the relisting period and threshold
+func (e *EventedPLEG) Update(relistDuration *RelistDuration) {
+	e.genericPleg.Update(relistDuration)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more
+	// than the relistThreshold. In case EventedPLEG is turned on,
+	// relistingPeriod and relistingThreshold are adjusted to higher
+	// values. So the health check of Generic PLEG should check
+	// the adjusted values of relistingPeriod and relistingThreshold.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) == cap(e.eventChannel) {
+		return false, fmt.Errorf("EventedPLEG: pleg event channel capacity is full with %v events", len(e.eventChannel))
+	}
+
+	timestamp := e.clock.Now()
+	metrics.PLEGLastSeen.Set(float64(timestamp.Unix()))
+	return true, nil
+}
+
+func (e *EventedPLEG) watchEventsChannel() {
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
+	defer close(containerEventsResponseCh)
+
+	// Get the container events from the runtime.
+	go func() {
+		numAttempts := 0
+		for {
+			if numAttempts >= e.eventedPlegMaxStreamRetries {
+				if isEventedPLEGInUse() {
+					// Fall back to Generic PLEG relisting since Evented PLEG is not working.
+					klog.V(4).InfoS("Fall back to Generic PLEG relisting since Evented PLEG is not working")
+					e.Stop()
+					e.genericPleg.Stop()       // Stop the existing Generic PLEG which runs with longer relisting period when Evented PLEG is in use.
+					e.Update(e.relistDuration) // Update the relisting period to the default value for the Generic PLEG.
+					e.genericPleg.Start()
+					break
+				}
+			}
+
+			err := e.runtimeService.GetContainerEvents(containerEventsResponseCh)
+			if err != nil {
+				numAttempts++
+				e.Relist() // Force a relist to get the latest container and pods running metric.
+				klog.V(4).InfoS("Evented PLEG: Failed to get container events, retrying: ", "err", err)
+			}
+		}
+	}()
+
+	if isEventedPLEGInUse() {
+		e.processCRIEvents(containerEventsResponseCh)
+	}
+}
+
+func (e *EventedPLEG) processCRIEvents(containerEventsResponseCh chan *runtimeapi.ContainerEventResponse) {
+	for event := range containerEventsResponseCh {
+		podID := types.UID(event.PodSandboxStatus.Metadata.Uid)
+		shouldSendPLEGEvent := false
+
+		status, err := e.runtime.GeneratePodStatus(event)
+		if err != nil {
+			// nolint:logcheck // Not using the result of klog.V inside the
+			// if branch is okay, we just use it to determine whether the
+			// additional "podStatus" key and its value should be added.
+			if klog.V(6).Enabled() {
+				klog.ErrorS(err, "Evented PLEG: error generating pod status from the received event", "podUid", podID, "podStatus", status)
```

Comment:
nit: s/podUid/podUID here and below, most of the existing kubelet logs use `PodUID` (different casing), might be worth to stay consistent in case folks are using this to extract logs (https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/kubernetes/kubernetes%24+file:%5Epkg/kubelet+klog.*podUID.*&patternType=regexp)

Reply:
Done!